server/ntb/io/feeding_services/newsworthy.py

```diff
@@ -0,0 +1,166 @@
+# -*- coding: utf-8; -*-
+#
+# This file is part of Superdesk.
+#
+# Copyright 2013, 2014 Sourcefabric z.u. and contributors.
+#
+# For the full copyright and license information, please see the
+# AUTHORS and LICENSE files distributed with this source code, or
+# at https://www.sourcefabric.org/superdesk/license
+
+
+import logging
+import json
+import sys
+from os.path import join
+from hashlib import md5
+from tempfile import NamedTemporaryFile
+from flask import request
+import superdesk
+from superdesk.io.registry import register_feeding_service, register_feeding_service_parser
+from superdesk.io.feeding_services import FeedingService
+from superdesk.io.commands import update_ingest
+from superdesk.io import webhooks
+from superdesk.errors import IngestApiError, SuperdeskIngestError
+
+logger = logging.getLogger(__name__)
+EVENT_UNPUBLISHED = "newsitem.unpublished"
+
+
+class NewsworthyFeedingServiceAuth(webhooks.FeedingServiceWebhookAuth):
+    # FIXME: we can't use eve.auth.BasicAuth because eve uses flask.g.user
+    #        internally and superdesk.audit too, resulting in conflict
+
+    def authorized(self, allowed_roles, resource, method):
+        """Check that webhook data correspond to a known ingest provider
+
+        the provider is then registered in the service which will handle the update
+        """
+        username = request.authorization.get("username")
+        password = request.authorization.get("password")
+        lookup = {'feeding_service': NewsworthyFeedingService.NAME}
+
+        # if at least one provider match this request, we return True
+        found_provider = False
+
+        for provider in superdesk.get_resource_service('ingest_providers').get(req=None, lookup=lookup):
+            config = provider['config']
+            if config['username'] == username and config['password'] == password:
+                secret = (config['secret'] or '').encode('utf-8')
+                checksum = md5(json.dumps(request.json).encode('utf-8') + secret).hexdigest()
+
+                if sys.version_info[:2] <= (3, 5):
+                    # dict was not keeping order before Python 3.6, so the checksum would be random
```

Comment:
can we use https://docs.python.org/3/library/collections.html#collections.OrderedDict in this case?

Reply:
@ride90 I've thought about that too, but it would overcomplicate (to do the json parsing in order) while it's not something super important (the identification is done with user/password using HTTP basic auth, checksum is just another security). I don't think it worths it, for a version will ditch sooner or later (the sooner the better).